src/bluez/adapter.rs

```diff
@@ -0,0 +1,129 @@
+            devices
+                .into_iter()
+                .map(|device| CentralEvent::DeviceDiscovered(BDAddr::from(&device.mac_address))),
+        );
+
+        let session = self.session.clone();
+        let events = events.filter_map(move |event| central_event(event, session.clone()));
+
+        Ok(Box::pin(initial_events.chain(events)))
+    }
+
+    async fn start_scan(&self) -> Result<()> {
+        self.session.start_discovery().await?;
+        Ok(())
+    }
+
+    async fn stop_scan(&self) -> Result<()> {
+        self.session.stop_discovery().await?;
+        Ok(())
+    }
+
+    async fn peripherals(&self) -> Result<Vec<Peripheral>> {
+        let devices = self.session.get_devices().await?;
+        Ok(devices
+            .into_iter()
+            .map(|device| Peripheral::new(self.session.clone(), device))
+            .collect())
+    }
+
+    async fn peripheral(&self, address: BDAddr) -> Result<Peripheral> {
+        let devices = self.session.get_devices().await?;
+        devices
+            .into_iter()
+            .find_map(|device| {
+                if BDAddr::from(&device.mac_address) == address {
+                    Some(Peripheral::new(self.session.clone(), device))
+                } else {
+                    None
+                }
+            })
+            .ok_or(Error::DeviceNotFound)
+    }
+}
+
+impl From<BluetoothError> for Error {
+    fn from(error: BluetoothError) -> Self {
+        Error::Other(error.to_string())
+    }
+}
+
+async fn central_event(event: BluetoothEvent, session: BluetoothSession) -> Option<CentralEvent> {
+    match event {
+        BluetoothEvent::Device {
+            id,
+            event: DeviceEvent::Discovered,
+        } => {
+            let device = session.get_device_info(&id).await.ok()?;
+            Some(CentralEvent::DeviceDiscovered((&device.mac_address).into()))
+        }
+        BluetoothEvent::Device {
+            id,
+            event: DeviceEvent::Connected { connected },
+        } => {
+            let device = session.get_device_info(&id).await.ok()?;
+            if connected {
+                Some(CentralEvent::DeviceConnected((&device.mac_address).into()))
+            } else {
+                Some(CentralEvent::DeviceDisconnected(
+                    (&device.mac_address).into(),
+                ))
+            }
+        }
+        BluetoothEvent::Device {
+            id,
+            event: DeviceEvent::RSSI { rssi: _ },
+        } => {
+            let device = session.get_device_info(&id).await.ok()?;
+            Some(CentralEvent::DeviceUpdated((&device.mac_address).into()))
+        }
+        BluetoothEvent::Device {
+            id,
+            event: DeviceEvent::ManufacturerData { manufacturer_data },
+        } => {
+            let device = session.get_device_info(&id).await.ok()?;
+            if manufacturer_data.len() > 1 {
+                warn!("Got more than one manufacturer data entry.")
+            }
```

Comment:
Would it be a better idea to emit an event for each manufacturer data that is received, instead of dropping them?

Reply:
Rebased on your PR, which solves this problem in a better way.